xpdan/search.py

```diff
@@ -0,0 +1,358 @@
+                        (max(internal_scores), h['start']['time'] * -1, h))
+        heap.sort()
+        heap.reverse()
+        return [g[-1] for g in heap if g[0] != -1]
+
+    def beamtime_dates(self, keys=('beamtime_uid', 'bt_safN',
+                                   'facility', 'beamline'),
+                       beamtime_key='beamtime_uid',
+                       print_results=True):
+        """Get info for each beamtime
+
+        Parameters
+        ----------
+        keys: iterable of str
+            The keys to be included in the return
+        beamtime_key: str
+            The key for the unique beamtime key
+        print_results: bool
+            If true prints the information
+
+        Returns
+        -------
+        list of dicts:
+            The list of beamtimes and their associated information
+        """
+        hdrs = self()
+        bts = set([h['start'][beamtime_key] for h in hdrs])
+        returns = []
+        for s in bts:
+            hdrs = self(**{beamtime_key: s})
+            start_hdr = hdrs[0]
+            stop_hdr = hdrs[-1]
+            info = {k: start_hdr[k] for k in keys if k in start_hdr.keys()}
+            info.update({'start_time': _munge_time(start_hdr['start']['time'],
+                                                   pytz.timezone(
+                                                       'US/Eastern')),
+                         'stop_time': _munge_time(stop_hdr['start']['time'],
+                                                  pytz.timezone(
+                                                      'US/Eastern'))})
+            returns.append(info)
+        if print_results:
+            pprint(returns)
+        return returns
+
+    def fuzzy_set_search(self, key, search_string, size=100):
+        """Return the most similar set of values to the search string.
+
+        Parameters
+        ----------
+        key: list of str
+            The list of strings to be accessed
+        search_string: str
+            The string to be searched for
+        size: int, optional
+            The number of results to be returned.
+             Defaults to 100 results
+
+        Returns
+        -------
+        list:
+            A list of headers which are close to the query. The queried value
+            will be unique in the list so a search for piLast='Alice' will
+            return only one header with Alice as the PI.
+
+        Examples
+        --------
+        >>> db = Broker(...) # Contains runs from Bob, Alice, Bob, and Eve
+        >>> fuzzy_set_search(db, 'bt_piLast', 'Bob')
+        ['Bob', 'Alice', 'Eve']
+        """
+        heap = [(-1, -1)] * size  # ndld can't return less than 0
+        heapify(heap)
+        values = set([h['start'][key] for h in self()])
+        for v in values:
+            heappushpop(heap, (1. - ndld(v, search_string), v))
+        heap.sort()
+        heap.reverse()
+        return [g[-1] for g in heap if g[0] >= 0.]
+
+
+try:
```

Comment:
Don't do this, we have learned this is a bad pattern.  If you want to have an auto-configure module make a new module _just_ for the singleton.